test/FunctionalTests/Microsoft.OData.Client.Tests/ALinq/PreserveTypesAndAttributesTests.cs

```diff
@@ -0,0 +1,395 @@
+    }
+
+    [Key("CustomerID")]
+    public class OneLevelCustomer1 : ICustomer
+    {
+        [OriginalName("CustomerID")]
+        public string Id { get; set; }
+
+        public string City { get; set; }
+
+        [OriginalName("CompanyName")]
+        public string CorporationName { get; set; }
+
+        [OriginalName("ContactName")]
+        public string Name { get; set; }
+    }
+
+    [Key("CustomerID")]
+    public class BaseCustomer
+    {
+        [OriginalName("CustomerID")]
+        public string Id { get; set; }
+    }
+
+    public class OneLevelCustomer2 : BaseCustomer
+    {
+        public string City { get; set; }
+
+        [OriginalName("CompanyName")]
+        public string CorporationName { get; set; }
+
+        [OriginalName("ContactName")]
+        public string Name { get; set; }
+    }
+
+
+    // More than one level of inheritance classes
+
+    public abstract class ACustomer : ICustomer
+    {
+        public virtual string Id { get; set; }
+        public virtual string City { get; set; }
+        public virtual string CorporationName { get; set; }
+        public virtual string Name { get; set; }
+    }
+
+    [Key("CustomerID")]
+    public class TwoLevelCustomer1 : ACustomer
+    {
+        [OriginalName("CustomerID")]
+        public override string Id { get; set; }
+
+        public override string City { get; set; }
+
+        [OriginalName("CompanyName")]
+        public override string CorporationName { get; set; }
+
+        [OriginalName("ContactName")]
+        public override string Name { get; set; }
+    }
+
+    public interface ICustomer2 : ICustomer
+    {
+        List<Order> OrdersList { get; set; }
+    }
+
+    [Key("CustomerID")]
+    public class TwoLevelCustomer2 : ICustomer2
+    {
+        [OriginalName("CustomerID")]
+        public string Id { get; set; }
+
+        public string City { get; set; }
+
+        [OriginalName("CompanyName")]
+        public string CorporationName { get; set; }
+
+        [OriginalName("ContactName")]
+        public string Name { get; set; }
+
+        [OriginalName("Orders")]
+        public List<Order> OrdersList { get; set; }
+    }
+
+    [Key("OrderID")]
+    public class Order
+    {
+        public int OrderID { get; set; }
+        public decimal Freight { get; set; }
+    }
+
+
+    public class TestContext
+    {
+        public DataServiceQuery<OneLevelCustomer1> OneLevelCustomers1;
+        public DataServiceQuery<TwoLevelCustomer1> TwoLevelCustomers1;
+        public DataServiceQuery<OneLevelCustomer2> OneLevelCustomers2;
+        public DataServiceQuery<TwoLevelCustomer2> TwoLevelCustomers2;
+
+        // We have to create multiple DataServiceContext because the entitySetName is the same for every DataServiceQuery
+        private readonly DataServiceContext _ctx1;
+        private readonly DataServiceContext _ctx2;
+        private readonly DataServiceContext _ctx3;
+        private readonly DataServiceContext _ctx4;
```

Comment:
I don't think it's necessary to have 4 different instances. Since this `TestContext` class is instantiated in the constructor of the test class, it means it's instantiated for each test. Each test seems to be using only one of the `DataServiceContext` instance, but all 4 are created. So, for each test, we create 3 instances that we don't use, which is wasteful. I would suggest refactoring the tests such that we only create one instance. Do reduce boilerplate inside the test methods, this instance can be initialized in the test class constructor. Since we're not using static fields, we shouldn't have to worry about one test method interfering with the state of a different test method.